File: 837. New 21 Game/optimised.py
#!/usr/bin/env python

def new21Game(n: int, k: int, maxPts: int) -> float:
    if k == 0 or n >= k + maxPts:
        return 1.0

    dp = [0] * (k + maxPts)

    for i in range(k, n + 1):
        dp[i] = 1

    windowSum = 0

    for i in range(k, k + maxPts):
        windowSum += dp[i]

    for i in range(k - 1, -1, -1):
        dp[i] = windowSum / maxPts
        windowSum += dp[i] - dp[i + maxPts]

    return dp[0]

print(new21Game(10, 1, 10))

File: 837. New 21 Game/optimised.ts
function new21Game(n: number, k: number, maxPts: number): number {
	if (k == 0 || n >= k + maxPts) {
		return 1.0;
	}

	const dp = new Array(k + maxPts).fill(0);

	for (let i = k; i < n + 1; i++) {
		dp[i] = 1;
	}

	let wSum = 0;

	for (let i = k; i < k + maxPts; i++) {
		wSum += dp[i];
	}

	for (let i = k - 1; i > -1; i--) {
		dp[i] = wSum / maxPts;
		wSum += dp[i] - dp[i + maxPts];
	}

	return dp[0];
};

console.log(new21Game(10, 1, 10));
